protocol.go

```diff
@@ -0,0 +1,124 @@
+package mssql
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/microsoft/go-mssqldb/msdsn"
+)
+
+// ProtocolDialer makes the network connection for a protocol
+type ProtocolDialer interface {
+	// Translates data from SQL Browser to parameters in the config
+	ParseBrowserData(data BrowserData, p *msdsn.Config) error
+	// Returns a Dialer to make the connection
+	DialConnection(ctx context.Context, c *Connector, p msdsn.Config) (conn net.Conn, err error)
+	// Returns true if information is needed from the SQL Browser service to make a connection
+	CallBrowser(p *msdsn.Config) bool
+}
+
+// ProtocolResolvers should be updated by any package that provides a ProtocolResolver and an associated ProtocolDialer
+var ProtocolDialers map[string]ProtocolDialer = map[string]ProtocolDialer{
```

Comment:
I may have to move some of these data types into a separate package to avoid an import cycle in the tests.

The tests are in the mssql package, so they can't import a "namedpipe" package that also imports mssql.

Reply:
i could define a version of ProtocolDialer inside msdsn that doesn't accept a Connector parameter, and a TcpProtocolDialer here that accepts the Connector. 
The npipe dialer doesn't need a Connector, as it puts all its needed data in the Config.
Or I could try stashing the Connector instance in the Config as part of the tcp properties.